File: lua/test/git_ref_entity_test.lua
-- GitRef entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("github_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("GitRefEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:GitRef(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = git_ref_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "git_ref." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_GIT_REF_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local git_ref_ref01_ent = client:GitRef(nil)
    local git_ref_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.git_ref"), "git_ref_ref01"))
    git_ref_ref01_data["owner"] = setup.idmap["owner01"]
    git_ref_ref01_data["repo"] = setup.idmap["repo01"]

    local git_ref_ref01_data_result, err = git_ref_ref01_ent:create(git_ref_ref01_data, nil)
    assert.is_nil(err)
    git_ref_ref01_data = helpers.to_map(type(git_ref_ref01_data_result) == 'table' and git_ref_ref01_data_result.data_get and git_ref_ref01_data_result:data_get() or git_ref_ref01_data_result)
    assert.is_not_nil(git_ref_ref01_data)
    assert.is_not_nil(git_ref_ref01_data["id"])

    -- UPDATE
    local git_ref_ref01_data_up0_up = {
      id = git_ref_ref01_data["id"],
      ["owner"] = setup.idmap["owner"],
      ["repo"] = setup.idmap["repo"],
    }

    local git_ref_ref01_markdef_up0_name = "node_id"
    local git_ref_ref01_markdef_up0_value = "Mark01-git_ref_ref01_" .. tostring(setup.now)
    git_ref_ref01_data_up0_up[git_ref_ref01_markdef_up0_name] = git_ref_ref01_markdef_up0_value

    local git_ref_ref01_resdata_up0_result, err = git_ref_ref01_ent:update(git_ref_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local git_ref_ref01_resdata_up0 = helpers.to_map(type(git_ref_ref01_resdata_up0_result) == 'table' and git_ref_ref01_resdata_up0_result.data_get and git_ref_ref01_resdata_up0_result:data_get() or git_ref_ref01_resdata_up0_result)
    assert.is_not_nil(git_ref_ref01_resdata_up0)
    assert.are.equal(git_ref_ref01_resdata_up0["id"], git_ref_ref01_data_up0_up["id"])
    assert.are.equal(git_ref_ref01_resdata_up0[git_ref_ref01_markdef_up0_name], git_ref_ref01_markdef_up0_value)

    -- LOAD
    local git_ref_ref01_match_dt0 = {
      id = git_ref_ref01_data["id"],
    }
    local git_ref_ref01_data_dt0_loaded, err = git_ref_ref01_ent:load(git_ref_ref01_match_dt0, nil)
    assert.is_nil(err)
    local git_ref_ref01_data_dt0_load_result = helpers.to_map(type(git_ref_ref01_data_dt0_loaded) == 'table' and git_ref_ref01_data_dt0_loaded.data_get and git_ref_ref01_data_dt0_loaded:data_get() or git_ref_ref01_data_dt0_loaded)
    assert.is_not_nil(git_ref_ref01_data_dt0_load_result)
    assert.are.equal(git_ref_ref01_data_dt0_load_result["id"], git_ref_ref01_data["id"])

  end)
end)

function git_ref_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/git_ref/GitRefTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read git_ref test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "git_ref01", "git_ref02", "git_ref03", "repo01", "repo02", "repo03", "matching_ref01", "matching_ref02", "matching_ref03", "owner01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITHUB_TEST_GIT_REF_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITHUB_TEST_GIT_REF_ENTID"] = idmap,
    ["GITHUB_TEST_LIVE"] = "FALSE",
    ["GITHUB_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["GITHUB_TEST_GIT_REF_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["owner"] == nil then
    idmap_resolved["owner"] = idmap_resolved["owner01"]
  end
  if idmap_resolved["repo"] == nil then
    idmap_resolved["repo"] = idmap_resolved["repo01"]
  end

  if env["GITHUB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITHUB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITHUB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/git_ref_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestGitRefEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.GitRef(nil)
		if ent == nil {
			t.Fatal("expected non-nil GitRefEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := git_refBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "git_ref." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_GIT_REF_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		gitRefRef01Ent := client.GitRef(nil)
		gitRefRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "git_ref"}), "git_ref_ref01"))
		gitRefRef01Data["owner"] = setup.idmap["owner01"]
		gitRefRef01Data["repo"] = setup.idmap["repo01"]

		gitRefRef01DataResult, err := gitRefRef01Ent.Create(gitRefRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		gitRefRef01Data = core.ToMapAny(entityData(gitRefRef01DataResult))
		if gitRefRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if gitRefRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// UPDATE
		gitRefRef01DataUp0Up := map[string]any{
			"id": gitRefRef01Data["id"],
			"owner": setup.idmap["owner"],
			"repo": setup.idmap["repo"],
		}

		gitRefRef01MarkdefUp0Name := "node_id"
		gitRefRef01MarkdefUp0Value := fmt.Sprintf("Mark01-git_ref_ref01_%d", setup.now)
		gitRefRef01DataUp0Up[gitRefRef01MarkdefUp0Name] = gitRefRef01MarkdefUp0Value

		gitRefRef01ResdataUp0Result, err := gitRefRef01Ent.Update(gitRefRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		gitRefRef01ResdataUp0 := core.ToMapAny(entityData(gitRefRef01ResdataUp0Result))
		if gitRefRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if gitRefRef01ResdataUp0["id"] != gitRefRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if gitRefRef01ResdataUp0[gitRefRef01MarkdefUp0Name] != gitRefRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", gitRefRef01MarkdefUp0Name, gitRefRef01ResdataUp0[gitRefRef01MarkdefUp0Name])
		}

		// LOAD
		gitRefRef01MatchDt0 := map[string]any{
			"id": gitRefRef01Data["id"],
		}
		gitRefRef01DataDt0Loaded, err := gitRefRef01Ent.Load(gitRefRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		gitRefRef01DataDt0LoadResult := core.ToMapAny(entityData(gitRefRef01DataDt0Loaded))
		if gitRefRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if gitRefRef01DataDt0LoadResult["id"] != gitRefRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func git_refBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "git_ref", "GitRefTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read git_ref test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse git_ref test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"git_ref01", "git_ref02", "git_ref03", "repo01", "repo02", "repo03", "matching_ref01", "matching_ref02", "matching_ref03", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_GIT_REF_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_GIT_REF_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_GIT_REF_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add owner alias for update test.
	if idmapResolved["owner"] == nil {
		idmapResolved["owner"] = idmapResolved["owner01"]
	}
	// Add repo alias for update test.
	if idmapResolved["repo"] == nil {
		idmapResolved["repo"] = idmapResolved["repo01"]
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
